### PR TITLE
Wip json parsing stage tweaks

### PR DESCRIPTION
**Pull request by fsat**
@ktoso - minor tweaks to JSON streaming:
- Create less `ByteString`
- Do not mutate characters between JSON object (i.e. `{` and `}`)

I can add further modifications as mentioned inside the code's TODO list

### CODE REVIEW

**Comment by fsat** [comment]
@ktoso - I've some questions as well:
- Shall I rename `JsonCollectingBuffer` to `JsonFrameCounting`?
- Is it really possible to use just one `Int` to keep track of `{` and `}` since the streaming push and pull can move independently of each other (i.e. 3 objects pushed vs 1 pulled)?
- If you remove default value of `maxObjectLength`, should this value come from the config or user specified as the request `stream[MyObject]`?

File: akka-stream-tests/src/test/scala/akka/stream/io/JsonFramingSpec.scala [hunk 1]
```diff
@@ -38,9 +38,9 @@ class JsonFramingSpec extends AkkaSpec with ScalaFutures {
         }
 
       result.futureValue shouldBe Seq(
-        """{"name":"john"}""",
-        """{"name":"jack"}""",
-        """{"name":"katie"}""")
+        """{ "name": "john" }""".stripMargin,
+        """{ "name": "jack" }""".stripMargin,
+        """{ "name": "katie" }""".stripMargin)
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/io/JsonFramingSpec.scala:43]: `stripMargin` not really needed here?

File: akka-stream-tests/src/test/scala/akka/stream/io/JsonFramingSpec.scala [hunk 2]
```diff
@@ -101,8 +100,9 @@ class JsonFramingSpec extends AkkaSpec with ScalaFutures {
         }
 
       result.futureValue shouldBe Seq(
-        """{"name":"john"}""",
-        """{"name":"jack"}""")
+        """{ "name": "john"
+          |}""".stripMargin,
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/io/JsonFramingSpec.scala:104]:
```
        """|{ "name": "john"
           |}""".stripMargin,
```

?

File: akka-stream-tests/src/test/scala/akka/stream/io/JsonFramingSpec.scala [hunk 3]
```diff
@@ -193,7 +193,13 @@ class JsonFramingSpec extends AkkaSpec with ScalaFutures {
               |   }
               |}
               | """.stripMargin))
-          buffer.poll().get.get.utf8String shouldBe """{"name":"john","age":101,"address":{"street":"Straight Street","postcode":1234}}"""
+          buffer.poll().get.get.utf8String shouldBe """{  "name": "john",
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/io/JsonFramingSpec.scala:196]: add | in first line as well, easier to eye-tray things :)

File: akka-stream/src/main/scala/akka/stream/io/JsonCollectingBuffer.scala [hunk 4]
```diff
@@ -71,8 +71,8 @@ class JsonCollectingBuffer(maximumObjectLength: Int = Int.MaxValue) {
       val possibleResult = completedObjectIndexes.headOption
       if (possibleResult.isDefined) {
         val index = possibleResult.get
-        val result = buffer.slice(0, index)
-        buffer = buffer.slice(index, buffer.length)
+        val (result, remainder) = buffer.splitAt(index)
+        buffer = remainder
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/io/JsonCollectingBuffer.scala:75]: good :)
Might benefit from calling `.compact()`? // cc @drewhk

File: akka-stream/src/main/scala/akka/stream/io/JsonCollectingBuffer.scala [hunk 5]
```diff
@@ -93,28 +93,30 @@ class JsonCollectingBuffer(maximumObjectLength: Int = Int.MaxValue) {
...
     } else if (input == CurlyBraceStart && !isStartOfStringExpression) {
       isStartOfEscapeSequence = false
       objectDepthLevel += 1
-      buffer ++= ByteString(input)
+      buffer :+= input
     } else if (input == CurlyBraceEnd && !isStartOfStringExpression) {
       isStartOfEscapeSequence = false
       objectDepthLevel -= 1
-      buffer ++= ByteString(input)
+      buffer :+= input
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/io/JsonCollectingBuffer.scala:108]: :+1:

**Comment by ktoso** [comment]
Looks like good changes!

> Shall I rename JsonCollectingBuffer to JsonFrameCounting?

Or maybe `JsonBraceCountingFramer` - Not sure if too silly name :)

> Is it really possible to use just one Int to keep track of { and } since the streaming push and pull can move independently of each other (i.e. 3 objects pushed vs 1 pulled)?

I believe so, yes:
- pushed bytestring is not complete object (`pos = -1`) => we call `ctx.pull()` right away (in onPush)
- pushed bytestring has complete object (`pos = n`) => we push it
- pushed bytestring has more objects (`pos = n`, and after polling `pos` is still `> -1`) => we can push right away, and in onPull we can `if(buf.peek().isDefined) ctx.push(buf.poll()) else ctx.pull()` (something like that)
  - when `pos = NoIdea`, scan the bytestring for an element
  - if not complete yet `pos = ScannedTo; gotElement = false`; push comes in, we scan from `pos` until we have an element => `pos = n; gotElement = true`
  - if complete element found in the byte string `pos = n; gotElement = true`
  - `poll() called` – reset the gotElement
  - `poll() called again` – we need to scan for the element, update the pos and emit if gotElement - same alg as above :)

> If you remove default value of maxObjectLength, should this value come from the config or user specified as the request stream[MyObject]?

From config, in some way. Didn't yet clean-up the settings things :)

**Comment by ktoso** [comment]
Thanks for keeping at it!

**Comment by ktoso** [comment]
I've merged right away and next improvements can come once you're ready :)

**Pull request merged**